src/macros/src/lib.rs

```diff
@@ -0,0 +1,72 @@
+#![recursion_limit = "192"]
+
+extern crate proc_macro;
+extern crate syn;
+#[macro_use] extern crate quote;
+
+use proc_macro::TokenStream;
+
+
+#[proc_macro_derive(GfxVertexFormat)]
+pub fn vertex(input: TokenStream) -> TokenStream {
+    let s = input.to_string();
+    let ast = syn::parse_macro_input(&s).unwrap();
+    let gen = structure(ast, quote!(gfx::format::Formatted), quote!(gfx::format::Format));
+    gen.parse().unwrap()
+}
+
+#[proc_macro_derive(GfxConstantBuffer)]
+pub fn constant(input: TokenStream) -> TokenStream {
+    let s = input.to_string();
+    let ast = syn::parse_macro_input(&s).unwrap();
+    let gen = structure(ast, quote!(gfx::shade::Formatted), quote!(gfx::shade::ConstFormat));
+    gen.parse().unwrap()
+}
+
+fn structure(ast: syn::DeriveInput, ty_compile: quote::Tokens, ty_run: quote::Tokens)
+             -> quote::Tokens {
+    let name = &ast.ident;
+    let fields = match ast.body {
+        syn::Body::Struct(syn::VariantData::Struct(ref fields)) => fields,
+        _ => panic!("`GfxStruct` can only be derived for structs"),
```

Comment:
We are not really deriving `GfxStruct`, I think rustc already wraps the derive panics inside a pretty message with the derive name and location.